cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -149,9 +150,22 @@ class ParquetScanTaskIterator {
   static Status InferColumnProjection(const parquet::FileMetaData& metadata,
                                       const std::shared_ptr<ScanOptions>& options,
                                       std::vector<int>* out) {
-    // TODO(fsaintjacques): Compute intersection _and_ validity
-    *out = internal::Iota(metadata.num_columns());
+    if (options->projector == nullptr || options->schema == nullptr) {
+      *out = internal::Iota(metadata.num_columns());
+      return Status::OK();
+    }
 
+    // union fields in expression and fields in projector
+    auto projector_schema = options->projector->schema();
+    auto filter_schema =
+        SchemaFromColumnNames(options->schema, FieldsInExpression(options->filter));
```

Comment:
I don't think we should "merge" the schema with validation, the validation is already done in layers above. This should only make sure that the referenced column names in the filter are found in `options->schema`, if not, append. to it.

```
std::vector<std::string> fields_in_schema = options->schema->field_names();
std::vector<std::string> fields_in_filter = FieldsInExpression(...);

for (auto missing_field: set(fields_in_filter) - set(fields_in_schema)) {
  fields_in_schema.append(missing_field);
}

// Use fields_in_schema for the indices resolution loop.
```

Reply:
It seems you're asking for `MergeSchemas` to be inlined. I can do that but it seemed useful enough to be reused later. I don't think a redundant validation will hurt performance here